### DISCUSSION TITLE
using dosbox-staging for smooth scrolling (Asterix)

### DISCUSSION THREAD

**Discussion by littbarski** [dosbox-staging/dosbox-staging#3986] [Q&A]
Hello, 

I enjoy playing old DOS games and have managed to get (after years of trying) quite smooth scrolling in 2D games like jump n runs and top down or side shooter (Chaos Engine, Kelloggs and Nesquik game, Golden Axe). Until now I use DOSBox-X as I can set a sync rate and especially a refresh rate for the game. But of course I am always searching for improvement, mainly for one game called "Asterix & Obelix" from 1996 by Infogrames, probably quite known. 

The image stuttering is extreme, also in DOSBox staging. So I hope here for any hints if this DOSBox has further options to get this smooth? What would be a starting conf file or configuration? Or is this not the kind of task that is seen as most important perhaps? 

Thank you!

**Comment by rderooy** [comment]
I tried it myself with the eXoDOS config file, and it seems to be fine. I don't notice any real flickering. But note that I'm using a display that can run at >70Hz. If you have a typical 60Hz display that could explain flickering.

Here is the eXoDOS config file for reference. I think just the `cycles=50000` setting may be important.

```
# Config file for Dosbox-0.74
[sdl]
fullscreen=false
fulldouble=false
fullresolution=0x0
windowresolution=1280x960
output=overlay
mapperfile=mapper.map
[dosbox]
machine=svga_s3
memsize=16
[render]
aspect=true
scaler=normal2x
[cpu]
core=auto
cputype=auto
cycles=50000
[mixer]
[midi]
[sblaster]
[gus]
[speaker]
[joystick]
[serial]
[dos]
[ipx]
[autoexec]
cd ..
cd ..
mount c .\eXoDOS\AsterixO
imgmount d .\eXoDOS\AsterixO\cd\obelix.cue -t cdrom
c:
@cd OBELIX
@cls
@OBELIX
exit
```

**Reply by littbarski** [reply]
thank you very much. The 50000 and other options don't change the stuttering on my side unfortunately. Of course I don#t expect the perfect solution for my hardware and system. But my guess was just that there is a certain fork or certain setting that is "specialised" for smooth scrolling. 

I use normal DOSBox, DOSBox staging and DOSBox-X, until now I don't use exoDOS. So which fork is used there? 
I wonder why I could make the scrolling smooth with direct3, syncrate 60 and refresh rate 60 in many DOS games (which probably also have in part 70Hz), but for Asterix and Sensible Soccer it's really not good. 

Perhaps this all is not a very specific topic for DOSBox staging, on the other hand I thought that smooth scrolling could also be of interest here.

**Reply by interloper98** [reply]
Your system sounds like it works best at 60 fps max, so this could help in  Asterix & Obelix - it plays OK at 60 fps (conf is for Staging 0.82 rc1 or `main` build):

```ini
[dosbox]
machine = svga_s3
memsize = 16
dos_rate = 60

[cpu]
cputype = pentium
cpu_cycles_protected = 35000

[sblaster]
sbtype = sb16

[gus]
gus = off

[midi]
mpu401 = off
mididevice = none

[speaker]
pcspeaker = off

[voodoo]
voodoo = off

[ipx]
ipx = off

[ethernet]
ne2000 = off

[autoexec]
cd ..
cd ..
mount c .\eXoDOS\AsterixO
imgmount d .\eXoDOS\AsterixO\cd\obelix.cue -t cdrom
c:
@cd OBELIX
@cls
@OBELIX
exit
```

(here's some horizontal movement to show background smoothness.. be sure to use a good video player that doesn't drop frames :-)

https://github.com/user-attachments/assets/8bb1c136-6eca-45a9-bf27-364912ee6bf0

**Comment by littbarski** [comment]
Thank you, really glad to see people helping here. Actually I see the typical "no vsync" screen artefacts by looking at your video, no matter which video player I use. 

I tried again both of the conf files above and have still extreme stuttering or screen problems. 
As I am new to DOSBOX staging: my program searches under appdata the standard (or creates it). So there are always two conf files that the program uses, the standard and in addition the custom? 
Which conf file should be there under appdata to work good with one of your both custom conf file?

**Reply by littbarski** [reply]
ok I think I understood, the custom conf just overwrites some things that should be different from the standard appdata created conf. I think this should be fine in my system.

I tried a lot of options incl vsync and output but no improvement. Would be interesting if anybody with a 60Hz monitor can play this without vsync problems.

PS: To explain it better, I get quite ok scrolling but this typical vsync problem, in the case of Asterix when the person walks to the right then there is a horizontal line of screen problem walking from the bottom to the up, just as it is when seeing old televisions filmed. See example with arrows.
![2](https://github.com/user-attachments/assets/1c710f24-74eb-4bff-a2af-ebcde4682c3f)

**Reply by johnnovak** [reply]
@littbarski Yeah, the game stutters/judders/tears like crazy, no matter what settings I've tried. Same results in DOSBox-X. I think the other guys who said "it's fine" aren't used to butter-smooth scrolling you'd get on a C64, Amiga, or even some DOS games (e.g. Pinball Dreams, Prehistorik 2, the game you listed in your top description, etc.) 😆 

I can run my monitor at 70 Hz, but no improvement. The game quite obviously tears at the _emulation_ level (as evidenced by the video recording as well, like you said), so this is not frame-presentation related at the host level.

This must be a VGA emulation inaccuracy that affects all DOSBoxes. Something is not quite right in the bowels of the VGA code... I'm afraid we'll need to leave it at that for the time being, few people dare to touch that code 😅

**Reply by littbarski** [reply]
thank you very much. This is at least good to know :).
As you say, the other games can be played really well. 
It would be interesting to see if Asterix has the same issue when playing it in a Windows system inside DOSBox. I will check this, at the moment I only have Windows 3.1 in the DOSBox, but Asterix need 95. I will have to search my old disks...

**Reply by rderooy** [reply]
The game does have some tearing, but I don't see "stuttering like crazy". But perhaps my threshold is different... But to me it seems to have more glitches when I lower my refresh rate to 60Hz.

I tried to record the gameplay, but the dosbox-staging built-in recorder always captures at 70Hz it seems, and an external recorder was dropping frames.

**Reply by littbarski** [reply]
After all if it even on 70Hz has this tearing and more or less stuttering, then it is just a game that I won't play at the end. Just as John said, it seems to be something that DOSBox in any version can not correct. I played this game as a child and have the CD from that time, and I remembered that the game was too difficult and the gameplay not perfect. So I thought I would give it a try after I discovered that I could make the above mentioned games really very smooth in scrolling. Asterix is really stuttering on my 60Hz, the other DOS games are not. 
I learned really a lot in this discussion and thank you all.

**Reply by interloper98** [reply]
**Tearing** is horizontal dis-joint content _inside a frame_, where as **smooth scrolling** is  seeing all the games' intended frames. Both problem can exist separately.

Tearing happens when a game updates the video memory at the same time the video card is  DAC'ing those lines out to the CRT (the draw period).

Smart games avoid changing video memory during the draw period, and instead wait until after the current frame is drawn but before the next frame starts - it's called the "back porch".  Games like Jazz Jackrabbit and Pinball Dreams do this.

Some games even setup a couple frames in VRAM and work on the next frame while the VGA card is drawing the prior frame. Then during the back porch period, the game only has to point the video card to new (now ready) frame in VRAM.

These smart games figure out when the back porch period happens by monitoring the VGA status register. You can use this patch:

```diff
diff --git a/src/hardware/vga_misc.cpp b/src/hardware/vga_misc.cpp
index 78bbb6571..a0eec105c 100644
--- a/src/hardware/vga_misc.cpp
+++ b/src/hardware/vga_misc.cpp
@@ -24,6 +24,7 @@
 #include "pic.h"
 #include "vga.h"
 #include <cmath>
+#include <format>
 
 void vga_write_p3d4(io_port_t port, io_val_t value, io_width_t);
 uint8_t vga_read_p3d4(io_port_t port, io_width_t);
@@ -54,6 +55,7 @@ uint8_t vga_read_p3da(io_port_t, io_width_t)
                        retval |= 1;
                }
        }
+       LOG_MSG("polling 0x3DA: %s", std::format("b{:08b}", retval).c_str() );
        return retval;
 }
```

And you'll see games like Jazz and Pinball Dreams checking the VGA status register (CRTC port 0x3DA) constantly during game play:

```text
2024-10-16 09:23:04.622 | polling 0x3DA: b00000101
2024-10-16 09:23:04.622 | polling 0x3DA: b00000101
2024-10-16 09:23:04.622 | polling 0x3DA: b00000101
2024-10-16 09:23:04.622 | polling 0x3DA: b00000101
2024-10-16 09:23:04.622 | polling 0x3DA: b00001101
2024-10-16 09:23:04.622 | polling 0x3DA: b00001101
2024-10-16 09:23:04.622 | polling 0x3DA: b00001101
2024-10-16 09:23:04.622 | polling 0x3DA: b00001101
2024-10-16 09:23:04.622 | polling 0x3DA: b00001101
2024-10-16 09:23:04.622 | polling 0x3DA: b00001101
2024-10-16 09:23:04.622 | polling 0x3DA: b00001101
2024-10-16 09:23:04.622 | polling 0x3DA: b00001101
2024-10-16 09:23:04.622 | polling 0x3DA: b00001101
2024-10-16 09:23:04.622 | polling 0x3DA: b00001101
```

---

Tearing games like Asterix & Obelix and Heretic pay no regard to this concern and write to the video memory either as fast as possible or using a coarse system timer, so you see the game's state transition (at time+0 and time+1) simultaneously in a given frame. The new state is typically above the tear, and the old state below the tear.

The difference between the VGA refresh rate and these tearing games' loop rate will make some harmonic.. so you'll see the tear position marching up or down across time.

Using the above patch, you'll see these tearing games don't bother polling the status register, either. So I have no doubt Asterix & Obelix would tear on real DOS hardware, too (just like Heretic does).

---

When it comes to smooth scrolling - this is a problem with the host system. Like @johnnovak mentioned, his system can display at 70 Hz, so he'll have smooth scrolling in Asterix & Obelix (but still torn because that's just how the game is written).  @littbarski you mentioned your system only runs at 60 Hz, so you'll additionally have non-smooth scrolling (because your system can't show you all 70 frames/s from the game).

DOS games ported from console platforms often don't have this back porch polling logic and instead use a system timer that gets them close to the original platform timing (50 Hz on PAL or 60 Hz on NTSC).

**Reply by interloper98** [reply]
Once you know that the smooth scrolling games poll for the back porch, then chances are you can change the `dos_rate` and the game will "do the right thing" and update without tearing, just like you figured out @littbarski.

For example, here's Jazz Jackrabbit at `dos_rate = 30` Hz, 45 Hz, and 60 Hz. You can inspect these videos frame-by-frame, and there are no torn frames :smile: 


https://github.com/user-attachments/assets/c3cd0b7f-7b5e-44e2-8935-fc47a98475e5


https://github.com/user-attachments/assets/6e2dc31d-66c2-42ca-9ea5-e5c7086346f1


https://github.com/user-attachments/assets/a9ade1d3-2695-4202-a661-7f385a113e23


So DOSBox definitely gets this right, provided the game is written right.

If you try this with Asterix, the frames are torn no matter what. That's just how it is!

(Word of caution: this is all just a discussion about frame tearing.. keep in mind some games might use the back porch as the game clock, so changing `dos_rate` might have side effects like changing game timing, affecting music tempo, etc.. )

**Reply by littbarski** [reply]
thank you very much again, as I am not a developer I don't know all about terminology and even less about code :).
I mentioned "scrolling" as the scrolling is the most important part of the game appearance, and the tearing (in addition to stuttering) appears while scrolling, so for me as a non expert it would be a part of it but of course I see the difference you explained.
The patch you mentioned is probably only an example or can I really somehow using it for these kind of games? 
Yes, I noticed some changes in music for example when I set the refresh rate in DOSBox-X at 60Hz (probably the same thing as the staging "dos_rate" ?) in Chaos Engine then I have to figure out about the cycles in terms of game speed and music.

**Reply by interloper98** [reply]
Yes @littbarski ; to use 60 Hz in DOSBox Staging it's the `dos_rate` setting.

```ini
[dosbox]
dos_rate = 60
```

There's also this really interesting article (if you have a fast PC!) https://github.com/dosbox-staging/dosbox-staging/wiki/High-refresh-rate-gaming

**Reply by johnnovak** [reply]
> Tearing is horizontal dis-joint content inside a frame, where as smooth scrolling is seeing all the games' intended frames. Both problem can exist separately.

Well, I kinda disagree with that 😄 In C64 and Amiga emulator circles smooth scrolling always means "just like on hardware". If it's tearing, it cannot be "smooth". "Smooth scrolling" means "without any artifacts that affect 100% hardware-like smooth scrolling"--it does not matter what exactly makes it not smooth.

Well, that's how the term has always been used in discussions about other emulators, so OP and me are on the same page 😄 

> So I have no doubt Asterix & Obelix would tear on real DOS hardware, too (just like Heretic does).

It seems unlikely for an 1996 DOS game to screw up smooth scrolling royally like this. Side-scrollers without 100% smooth scrolling are not enjoyable, and you could make that work even on a 286 with hardware VGA tricks (like Prehistorik 2). I'll test in on hardware today, but that won't be 100% proof either as I only have S3 cards which break such hardware tricks in some games... Oh well.

The rest of the technical explanation is spot on, yeah; that's how everybody synced screen updates to the vsync back then, including me. But this game must do it in an esoteric fashion, probably by getting the timing just right via other aspects of the VGA hardware that are not very well emulated. Definitely an outlier, though, I'll give you that. We'll see 😆

**Reply by interloper98** [reply]
> smooth scrolling always means "just like on hardware". If it's tearing, it cannot be "smooth"

Sadly no shortage of poorly codes DOS games tore on hardware. 70 Hz + torn frames was sometimes the best the machine could do 🤦‍♂️ 😂


> I'll test in on hardware today

Nice! Will be curious how it goes. 
👀

**Reply by interloper98** [reply]
Heretic was also tear happy, and it's a big budget AAA game (also proven not to poll for back porch just like Asterix).

**Reply by johnnovak** [reply]
> Some games even setup a couple frames in VRAM and work on the next frame while the VGA card is drawing the prior frame. Then during the back porch period, the game only has to point the video card to new (now ready) frame in VRAM.

Yeah, or on a faster CPU you could simply copy the backbuffer via the CPU. As long as you were a bit behind the position of the raster, you were golden. From fast 486s onwards people did not bother thst much with hardware VGA page flipping as that required Mode-X, they just started using it like a dumb framebuffer (e.g., all late DOS demos that require a Pentium).

**Reply by johnnovak** [reply]
> Heretic was also tear happy, and it's a big budget AAA game (also proven not to poll for back porch just like Asterix).

Yeah, it could be a crappily coded game 😆  Standards were just higher in Amiga circles 🤷🏻 I was apalled to see the jerky, jumpy nonsense in many DOS productions even in 1994, true... Maybe because they did not want to use unchained modes in those 3D games as that would slow down rendering due to the chunky to planar conversions, so they had to us standard chunky modes and let it tear.

Well, that made porting Doom to classic Amigas impossible; they only had planar (unchained) modes, like EGA and 640x480 16-colour standard VGA.

Maybe this game just has a low-effort DOS port as it can only run on Win95 too. Well, we'll see 😄

**Reply by littbarski** [reply]
hello, did you test it by the way :) ?

**Reply by johnnovak** [reply]
Nope, I did not.
